modules/integration/src/test/scala/scala/cli/integration/RunJdkTestDefinitions.scala

```diff
@@ -0,0 +1,133 @@
+    // TODO just passing the version number on arm64 should be enough, needs a fix in cs
+    if (Properties.isMac && TestUtil.isM1 && (javaVersion < 11 || javaVersion == 16))
+      s"zulu:$javaVersion"
+    else javaVersion.toString
+
+  def canUseScalaInstallationWrapper: Boolean =
+    actualScalaVersion.startsWith("3") && actualScalaVersion.split('.').drop(1).head.toInt >= 5
+
+  for {
+    javaVersion <- Constants.allJavaVersions
+    index = javaIndex(javaVersion)
+    useScalaInstallationWrapper <-
+      if (canUseScalaInstallationWrapper) Seq(false, true) else Seq(false)
+    launcherString = if (useScalaInstallationWrapper) "coursier scala installation" else "Scala CLI"
+    scalaRunnerWrapperVersion = actualScalaVersion match {
+      case v if v == Constants.scala3NextRc => Constants.scala3NextRcAnnounced
+      case v if v == Constants.scala3Next   => Constants.scala3NextAnnounced
+      case v                                => v
+    }
+    withLauncher = (root: os.Path) =>
+      (f: Seq[os.Shellable] => Unit) =>
+        if (useScalaInstallationWrapper)
+          withScalaRunnerWrapper(
+            root = root,
+            localBin = root / "local-bin",
+            scalaVersion = scalaRunnerWrapperVersion,
+            shouldCleanUp = false
+          )(launcher => f(Seq(launcher)))
+        else
+          f(Seq(TestUtil.cli))
+  } {
+    test(s"correct JVM is picked up by $launcherString when JAVA_HOME set to $index") {
+      TestUtil.retryOnCi() {
+        TestInputs(
+          os.rel / "check_java_home.sc" ->
+            s"""assert(
+               |  System.getProperty("java.version").startsWith("$javaVersion") ||
+               |  System.getProperty("java.version").startsWith("1.$javaVersion")
+               |)
+               |println(System.getProperty("java.home"))""".stripMargin
+        ).fromRoot { root =>
+          val javaHome =
+            os.Path(
+              os.proc(TestUtil.cs, "java-home", "--jvm", index).call().out.trim(),
+              os.pwd
+            )
+          withLauncher(root) { launcher =>
+            val res = os.proc(launcher, "run", ".", extraOptions)
+              .call(cwd = root, env = Map("JAVA_HOME" -> javaHome.toString))
+            expect(res.out.trim().contains(javaHome.toString))
+          }
+        }
+      }
+    }
+
+    test(s"hello world with $launcherString and --jvm $index") {
+      TestUtil.retryOnCi() {
+        val expectedMessage = "Hello, world!"
+        TestInputs(
+          os.rel / "hello_world.sc" -> s"println(\"$expectedMessage\")"
+        ).fromRoot { root =>
+          withLauncher(root) { launcher =>
+            val res = os.proc(launcher, "run", ".", extraOptions, "--jvm", javaVersion)
+              .call(cwd = root)
+            expect(res.out.trim() == expectedMessage)
+          }
+        }
+      }
+    }
+
+    if (!Properties.isWin || !useScalaInstallationWrapper) // TODO make this pass on Windows
```

Comment:
Why is not passing on windows?

Reply:
seemingly Scala CLI can't find the Java installation passed from `cs` via `--java-home`.
Since the same works when passed via `JAVA_HOME`, I assume the installation itself is valid.
It's likely that `--java-home` is broken on Windows, have to investigate separately.

Reply:
- https://github.com/VirtusLab/scala-cli/issues/3397: created an issue, let's track it there.